File: Java/10_LinkedList/LC_2130_MaximumTwinSumOfALinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }

        ListNode next = reverse(slow);

        int max = Integer.MIN_VALUE;
        while(next != null){
            max = Math.max(max,head.val + next.val);
            head = head.next;
            next = next.next;
        }

        return max;
    }

    private ListNode reverse(ListNode head){
        if (head == null || head.next == null) {
            return head;
        }
        ListNode pre = null;
        ListNode cur = head;
        while(cur != null){
            ListNode next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }

        return pre;
    }
}

public class LC_2130_MaximumTwinSumOfALinkedList {
    
}

File: Swift/Leetcode.playground/Sources/10_LinkedList/LC_2130_MaximumTwinSumOfALinkedList.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func pairSum(_ head: ListNode?) -> Int {

        var slow = head
        var fast = head
        
        while fast?.next != nil && fast?.next?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
        }

        let next = reverse(slow)
        var max = Int.min
        var currentHead = head
        var currentNext = next
        while currentHead != nil && currentNext != nil {
            guard let val1 = currentHead?.val, let val2 = currentNext?.val else{
                return 0
            }

            max = Swift.max(max, val1 + val2)
            currentHead = currentHead?.next
            currentNext = currentNext?.next
        }
        return max
    }

    func reverse(_ head: ListNode?) -> ListNode? {
        if head == nil || head?.next == nil {
            return head
        }
        var pre: ListNode? = nil
        var cur: ListNode? = head
        while cur != nil {
            var temp: ListNode? = cur?.next
            cur?.next = pre
            pre = cur
            cur = temp
        }
        return pre
    }
}